lib/clientv2/node.go

```diff
@@ -59,7 +63,103 @@ func (r nodes) Update(ctx context.Context, res *apiv2.Node, opts options.SetOpti
 
 // Delete takes name of the Node and deletes it. Returns an error if one occurs.
 func (r nodes) Delete(ctx context.Context, name string, opts options.DeleteOptions) (*apiv2.Node, error) {
+	pname, err := names.WorkloadEndpointIdentifiers{Node: name}.CalculateWorkloadEndpointName(true)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get all weps belonging to the node
+	weps, err := r.client.WorkloadEndpoints().List(ctx, options.ListOptions{
+		Prefix: true,
+		Name: pname,
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Collate all IPs across all endpoints, and then release those IPs.
+	ips := []net.IP{}
+	for _, wep := range weps.Items {
+		// The prefix match is unfortunately not a perfect match on the Node (since it is theoretically possible for
+		// another node to match the prefix (e.g. a node name of the format <thisnode>-foobar would also match a prefix
+		// search of the node <thisnode>). Therefore, we will also need to check that the Spec.Node field matches the Node.
+		if wep.Spec.Node != name {
+			continue
+		}
+		for _, ip := range wep.Spec.IPNetworks {
+			ips = append(ips, *net.ParseIP(ip))
+		}
+	}
+	_, err = r.client.IPAM().ReleaseIPs(context.Background(), ips)
+	switch err.(type) {
+	case nil, errors.ErrorResourceDoesNotExist, errors.ErrorOperationNotSupported:
+	default:
+		return nil, err
+	}
+
+	// Delete the weps.
+	for _, wep := range weps.Items {
+		if wep.Spec.Node != name {
+			continue
+		}
+
+		_, err = r.client.WorkloadEndpoints().Delete(ctx, wep.Namespace, wep.Name, options.DeleteOptions{})
+		switch err.(type) {
+		case nil, errors.ErrorResourceDoesNotExist, errors.ErrorOperationNotSupported:
+		default:
+			return nil, err
+		}
+	}
+
+	// Remove the node from the IPAM data if it exists.
+	err = r.client.IPAM().RemoveIPAMHost(ctx, name)
+	switch err.(type) {
+	case nil, errors.ErrorResourceDoesNotExist, errors.ErrorOperationNotSupported:
+	default:
+		return nil, err
+	}
+
+	// Remove BGPPeers.
+	bgpPeers, err := r.client.BGPPeers().List(ctx, options.ListOptions{})
+	if err != nil {
+		return nil, err
+	}
+	for _, peer := range bgpPeers.Items {
+		if peer.Spec.Node != name {
+			continue
+		}
+		_, err = r.client.BGPPeers().Delete(ctx, peer.Name, options.DeleteOptions{})
+		switch err.(type) {
+		case nil, errors.ErrorResourceDoesNotExist, errors.ErrorOperationNotSupported:
```

Comment:
Peers are supported on all datastore - so I don't think we should have the OperationNotSupported here as it'll potentially hide future bugs if we introduce another datastore.

Reply:
Same applies to BGPConfiguration and FelixConfiguration

Reply:
I'm confused - It will only hide `OperationNotSupported` errors.  Isn't that what we want?

Reply:
I'm saying that OperationNotSupported should never happen for BGPPeers, BGPConfig and FelixConfig on either KDD or etcdv3

We could leave it in place since it won't make any effective difference for now, but i don't really think we should be swallowing errors that we don't expect.

Reply:
ah leave it - perhaps it's safer in the long run

Reply:
I'll take the `git blame` when someday it comes back to bit us